hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/store/impl/MembershipStoreImpl.java

```diff
@@ -279,30 +279,30 @@ private MembershipState getRepresentativeQuorum(
       Collection<MembershipState> records) {
 
     // Collate objects by field value: field value -> order set of records
-    Map<FederationNamenodeServiceState, TreeSet<MembershipState>> occurenceMap =
+    Map<FederationNamenodeServiceState, Set<MembershipState>> occurenceMap =
         new HashMap<>();
     for (MembershipState record : records) {
       FederationNamenodeServiceState state = record.getState();
-      TreeSet<MembershipState> matchingSet = occurenceMap.get(state);
+      Set<MembershipState> matchingSet = occurenceMap.get(state);
       if (matchingSet == null) {
-        // TreeSet orders elements by descending date via comparators
-        matchingSet = new TreeSet<>();
+        matchingSet = new HashSet<>();
         occurenceMap.put(state, matchingSet);
       }
       matchingSet.add(record);
     }
 
     // Select largest group
-    TreeSet<MembershipState> largestSet = new TreeSet<>();
-    for (TreeSet<MembershipState> matchingSet : occurenceMap.values()) {
+    Set<MembershipState> largestSet = new HashSet<>();
```

Comment:
> This bug is because the above record is put into the Treeset in the method getRepresentativeQuorum. Since record1,2,3 have the same dateModified, there will only be one record in the final treeset of this method, so this method thinks that this nn is standby, because record4 newer

OK, so the issue is with the comparator and seems like going to HashSet, we fix the comparisson.
Shouldn't we be explicit about this instead of relying on TreeSet or HashSet using different comparators?
My proposal would be to fix that part.